### DISCUSSION TITLE
Overhead management for Kepler

### DISCUSSION THREAD

**Discussion by eklee15** [sustainable-computing-io/kepler#915] [General]
Questions/motivations 
* When collecting Kepler measurements from multiple nodes/pods, how much is the overhead?
  * How many nodes/pods can we scale out/up? (Nodes/pods)
  * What would be the requirements for scalability for Kepler?
* Targeted components
  * Prometheus Server
  * Kepler exporter
  * Grafana Dashboards

Related issues/discussions
* https://github.com/sustainable-computing-io/kepler/discussions/439
* https://github.com/sustainable-computing-io/kepler/discussions/836
* https://github.com/sustainable-computing-io/kepler/pull/823

**Comment by mcalman** [comment]
Thanks for opening this discussion @eklee15. I have begun some overhead/scalability testing using a local Kind cluster. So far I have been using [cAdvisor metrics](https://github.com/google/cadvisor/blob/master/docs/storage/prometheus.md) exposed by Prometheus to measure overhead. Here are some initial results:

**Scaling the size of a cluster with Kepler deployed**
I created a Kind cluster with Docker container "nodes". I then used [taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/) to deploy Kepler to a new node every 20 minutes and examined the overhead as the number of nodes with Kepler deployed increased. The plots below show the additional overhead consumed by the Prometheus container as Kepler is deployed on additional nodes.
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/5ad5b964-2fd1-46c9-ad30-aa49430742dd)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/b0352097-d1a5-4403-8856-ec37a7552b3a)

**Increasing the number of containers for Kepler to monitor**
For this experiment, every 5 minutes I added an additional dummy container (Debian containers executing integer addition every 10ms) to a single node with Kepler deployed.
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/ca3cfe6d-d272-4b22-9b99-e077aeca6a90)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/84bb3f74-b87a-48ae-926f-3f991641cbbe)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/ebadfd9b-3d3a-4343-9054-967ba7f98458)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/1a7fec3e-3087-4476-99cd-c48ae9b2007e)

**Reply by rootfs** [reply]
this is very helpful @mcalman 
the dummy container scalability test is very interesting, is the dummy container long running or transient? i suspect this could be related to the prometheus client library issue discussed [here](https://github.com/sustainable-computing-io/kepler/discussions/439)

**Reply by mcalman** [reply]
They are long-running. I created a deployment and then gradually scaled up the replicas

**Reply by rootfs** [reply]
thanks @mcalman, would it be possible to enable kepler cpu and mem profiler e.g. when there are 10 vs 40 replica?

**Reply by mcalman** [reply]
## CPU
**10 replicas**
`go tool pprof cpu.prof`
<img width="1109" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/3957a8d4-f613-404d-b580-842bb19b226d">
**40 replicas**
`go tool pprof cpu.prof`
<img width="1109" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/594ab6a2-1f25-4ec0-80a5-e463e65216c9">

## Memory
**10 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1109" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/fb5dadae-5d9b-435d-a1cc-0139bae6d13d">
**40 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1109" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/dcc1a524-f0a0-4fd6-a343-89a6d85aa0ed">

**Reply by rootfs** [reply]
@mcalman thanks, this is great! Would you please reformat the results using `top20 -cum (all)`? that'll sort the results in a better way.

**Reply by rootfs** [reply]
it looks `readdir` is a bottleneck in both 10 and 40 replica case, while `gcDrain` is more problematic in 40 replica sample.

**Reply by mcalman** [reply]
@rootfs thanks for the suggestion. I've updated the screenshots with the output of `top20 -cum (all)`.

**Reply by rootfs** [reply]
thank you @mcalman 

the `GetCgroupIDFromPath` call is interesting, if the cgroup ID and path are in the cache, this function shouldn't be called very often. Let's see what is going on.

**Reply by rootfs** [reply]
@mcalman sorry for one more request, do you have a chance to take a longer profiling? This time around the profiling appears to last `3020ms`. I remember your last profiling was longer than that. It could be possible the first `3020ms` were still ramping up the container cache. Once cache is stabilized, the profile will show something else. thanks!

**Reply by mcalman** [reply]
@rootfs sure I can do a longer profile. Just to clarify, this is the same profile as before (you should be able to see my previous screenshots in the edit history of my comment).

Each profile is 60s:
**10 Replica CPU Profile**: `Duration: 60s, Total samples = 3.02s ( 5.03%)`
**40 Replica CPU Profile**: `Duration: 60s, Total samples = 6.43s (10.72%)`

Would 5 minute profiles be worth trying?

**Reply by rootfs** [reply]
@mcalman I see, that make sense. If you can get 5 min profiles, that'll be great! I hope we can find the delta that can give us good pointers. thanks!

**Reply by mcalman** [reply]
## CPU
**10 replicas**
`go tool pprof cpu.prof`
`Duration: 300s, Total samples = 6.49s ( 2.16%)`
<img width="1253" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/236b0095-9171-4bf0-98ad-1407642e0522">
**40 replicas**
`go tool pprof cpu.prof`
`Duration: 300.10s, Total samples = 14.10s ( 4.70%)`
<img width="1253" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/7b5dabdf-9feb-405e-85da-362532ea0ff4">

## Memory
**10 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1251" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/e1288709-130e-4537-8a1f-de5578f35e4d">
**40 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1134" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/892b6c0e-555c-456f-bd1d-cb4607785146">

**Reply by rootfs** [reply]
thank you @mcalman for this quick results!

i think `cgroups/fs2.(*manager).GetStats` function is likely the one that causes all the scalability issue. Its call time is linear to the number of containers. 

Wonder if you can run yet another test with `EXPOSE_CGROUP_METRICS=true` to turn of cgroup stat? That'll probably be convincing to know if cgroup stat reader is indeed causing all the issues.

**Reply by marceloamaral** [reply]
`EXPOSE_CGROUP_METRICS=false`, right?

Actually, now that the ebpf metrics are fixed, we can probably turn off by default the cgroup metrics and also disable the kubelet metrics.

**Reply by rootfs** [reply]
right, `false` is to turn it off

@sunya-ch how are cgroup metric and kubelet models compared against the bpf models? is it ok to turn off cgroup metrics?

**Reply by mcalman** [reply]
# `EXPOSE_CGROUP_METRICS=false `
## CPU
**10 replicas**
`go tool pprof cpu.prof`
`Duration: 300s, Total samples = 1.94s ( 0.65%)`
<img width="1236" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/6dc10b25-6353-4b38-b59c-7f1c314b5a72">
**40 replicas**
`go tool pprof cpu.prof`
`Duration: 300s, Total samples = 3.13s ( 1.04%)`
<img width="1236" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/8aa32917-2937-47d1-8aa9-ce36b0246ae5">

## Memory
**10 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1109" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/a31c8fa2-b4b6-4d60-b932-20c358a80ce8">
**40 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1109" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/6f7914f8-2eca-4628-a175-1b21af26d26c">

**Reply by mcalman** [reply]
# `EXPOSE_CGROUP_METRICS=false EXPOSE_KUBELET_METRICS=false`
## CPU
**10 replicas**
`go tool pprof cpu.prof`
`Duration: 300s, Total samples = 1.94s ( 0.65%)`
<img width="1234" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/03cfcdb0-dad1-4dc2-8ab0-f9b2d3528fb9">
**40 replicas**
`go tool pprof cpu.prof`
`Duration: 300s, Total samples = 3.42s ( 1.14%)`
<img width="1234" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/26bd2e86-2958-4315-8bb5-01d033f1d904">

## Memory
**10 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1110" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/da75d351-8b55-4c10-8ed0-f36bc74b70bc">
**40 replicas**
`go tool pprof -alloc_objects mem.prof`
<img width="1110" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/beee7a09-f04c-40e8-8242-6f9904fe0a28">

**Reply by rootfs** [reply]
thank you @mcalman this result looks impressive!

It looks turning off cgroup stats has quite some impact on both 10 and 40 replicas, reducing cpu usage from 4.70% to 1.14%

Once we can confirm the bpf model precision from @sunya-ch we can use this as recommendations for deployment scenarios.

**Reply by rootfs** [reply]
actually @sunya-ch has published a comprehensive metrics on model accuracy [here](https://github.com/sustainable-computing-io/kepler-model-db/blob/main/models/Linux-4.15.0-213-generic-x86_64_v0.6/rapl_AbsPower_model_metadata.csv). It looks the cgroup models (MAE 188.0467120712669) are better than bpf models (MAE 267.55216545864715)

**Reply by sunya-ch** [reply]
@rootfs 
Sorry for delays but let me make an update. 
The pipeline [Linux-4.15.0-213-generic-x86_64_v0.6](https://github.com/sustainable-computing-io/kepler-model-db/tree/main/models/Linux-4.15.0-213-generic-x86_64_v0.6) is built before we have the BPF metric fixes. 
After fixes, we found the BPF CPU time is generally outperformed the current-exporting cgroup and kubelet metrics. Will update in detail with upcoming PR for kepler-model-db. 

Remark: there might be something wrong with the current cgroup and kubelet metrics though: check https://github.com/sustainable-computing-io/kepler/issues/901.

**Reply by rootfs** [reply]
thank you @sunya-ch ! this is a very good news!!

**Reply by mcalman** [reply]
> **Increasing the number of containers for Kepler to monitor**
> For this experiment, every 5 minutes I added an additional dummy container (Debian containers executing integer addition every 10ms) to a single node with Kepler deployed.

Here are the updated plots (scaling up to 80 dummy containers) I shared at the Kepler community meeting yesterday.
<img width="386" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/45a3046c-dae8-4967-86a1-30b98ddbc48c">
<img width="401" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/69dc54f1-c8ac-41a2-938d-ba96f1b2fd69">
<img width="401" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/c242cd29-5cba-45f4-9c02-9a28820ba4dc">
<img width="402" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/68564154/06558f03-6e5f-46b0-b77f-08f062456df2">

**Reply by rootfs** [reply]
@mcalman would it be possible to have your work hosted in this org so we can have more contributors?
wdyt @eklee15

**Reply by mcalman** [reply]
@rootfs Yes, I could do that.

**Reply by rootfs** [reply]
@mcalman great! invite sent to contribute to https://github.com/sustainable-computing-io/scalability-benchmark-suite

**Reply by marceloamaral** [reply]
These are great results, we can probably disable the cgroup and kubelet metrics by default.
@mcalman in summary, how much performance improvement did we get disabling the metrics for the dummy container test?

**Reply by mcalman** [reply]
~@marceloamaral For both 10 and 40 dummy containers the CPU utilization was almost exactly 2x greater with cgroup and kubelet metrics enabled. Additionally, network bytes transmitted per second by Kepler was 53x and 20x greater for 10 and 40 dummy containers respectively.~

Edit: For the "No cgroup/kubelet metrics" bars in these plots Prometheus wasn't scraping Kepler (the [release label was missing](https://github.com/sustainable-computing-io/kepler/issues/767#issuecomment-1717990301)) so these plots DO NOT show a fair comparison. See the updated plots [below](https://github.com/sustainable-computing-io/kepler/discussions/915#discussioncomment-7133806).

![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/98691a46-e042-4b63-abf8-f34eefdf1bd4)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/5348ed39-8fde-4ba8-9478-faed9294021b)

**Reply by eklee15** [reply]
Can you help run the overhead for the 1, 10, 20, 30, 40, 50, 60 ... 100 ... replicas until the CPU/network saturates? This result will be very interesting to see.

**Reply by rootfs** [reply]
it is good to see the reduced network traffic. I think fewer metrics helps reduce CPU in both getting cgroup and kubelet stats and generating these metrics. We should consider removing cgroup and kubelet in future releases when we get more model accuracy info.

**Reply by mcalman** [reply]
I reran the experiment scaling up to the max amount of pods I could deploy on a node (Kubernetes limit of 110 minus Prometheus and Kube system pods also present on the node. I also fixed an issue where Prometheus wasn't scraping Kepler metrics when I ran without kubelet and cgroup metrics the first time.

It looks like there isn't a huge reduction in network overhead, but CPU utilization is significantly reduced and scales better with the metrics excluded.
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/5c77ad97-7f92-4739-9296-066a1ba1bb63)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/a914f8d5-0c6a-4125-aede-957cdcd58fcb)

**Reply by rootfs** [reply]
thank you @mcalman we'll investigate what other metrics we can turn off to reduce network traffic

**Reply by marceloamaral** [reply]
@mcalman these are very good results. I think we can already disable the cgroup and kubelet metrics by default.

@rootfs increasing the Prometheus scrapping interval, from 3s to 30s will probably further minimize the overhead in Prometheus. Max has actually tested it already.

**Reply by mcalman** [reply]
@rootfs Here is the comparison @marceloamaral is referring to. The plots show Prometheus overhead that can be attributed to Kepler (I have subtracted out the average overhead values when Kepler is not deployed).

![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/80ffea30-6189-499c-bec1-a56890a5b2ea)
![image](https://github.com/sustainable-computing-io/kepler/assets/68564154/9359a7b1-1980-4b4e-b5d3-099b110abfbf)

**Reply by rootfs** [reply]
this is cool! @mcalman can you summarize your recommendations and add to the [project board](https://github.com/orgs/sustainable-computing-io/projects/2)

**Reply by mcalman** [reply]
@rootfs I believe I need to be granted access in order to add to the project board

**Reply by rootfs** [reply]
got it. Would you please create github issues and tasks? I'll link the project card to your issues/tasks

**Reply by mcalman** [reply]
Thanks @rootfs. Done https://github.com/sustainable-computing-io/kepler/issues/963#issue-1926808575

**Comment by husky-parul** [comment]
@AdrianHammond @sthaha @vimalk78

**Comment by eklee15** [comment]
@rootfs I think this experiment will be more interesting when we see the resource saturation. We are trying to see 3 different aspects. 
1) Number of dummy pod scaling (how many pods Kepler can cover if we increase the number of pods in BM/VM)
2) Number of nodes (vm) scaling (Kepler exporter can scale up to more than 100 or more nodes?) 
3) ebpf sampling rate. (what is the reasonable sampling rate?)

**Comment by rootfs** [comment]
let's take this research into action
https://github.com/sustainable-computing-io/kepler/issues/964
